In [1]:
from pathlib import Path
import re
import json
from typing import List, Dict, Optional

In [2]:
def find_repo_root(start: Optional[Path] = None, repo_name: str = "masters_thesis_sdg") -> Path:
    """
    Walk upward from the current working directory until the repository root is found.
    """
    if start is None:
        start = Path.cwd().resolve()

    current = start
    while True:
        if current.name == repo_name:
            return current
        if current.parent == current:
            raise FileNotFoundError(
                f"Could not find repo root '{repo_name}' starting from {start}"
            )
        current = current.parent


REPO_ROOT = find_repo_root()
DATA_DIR = REPO_ROOT / "data"
RAW_DIR = DATA_DIR / "raw" / "lai2023"
PROCESSED_DIR = DATA_DIR / "processed"

PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

# Fail loudly instead of silently producing an empty index: an rglob over a missing
# directory returns no files, which would otherwise look like a successful run of zero games.
if not RAW_DIR.is_dir():
    raise FileNotFoundError(
        f"Raw Lai et al. (2023) data not found at {RAW_DIR}. "
        "Download it first with src/dowload_lai.py."
    )

print("Repo root:", REPO_ROOT)
print("Raw dir:", RAW_DIR)
print("Processed dir:", PROCESSED_DIR)

Repo root: C:\Users\annab\Documents\GitHub\masters_thesis_sdg
Raw dir: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\raw\lai2023
Processed dir: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\processed


In [3]:
YT_TRANSCRIPTS_DIR = RAW_DIR / "Youtube" / "transcripts"
EGO4D_TRANSCRIPTS_DIR = RAW_DIR / "Ego4D" / "transcripts"

YT_OUTCOMES_DIR = RAW_DIR / "Youtube" / "vote_outcome_youtube_released"
EGO4D_OUTCOMES_DIR = RAW_DIR / "Ego4D" / "vote_outcome_ego4d"

OUTPUT_DIR = PROCESSED_DIR / "lai2023" / "onuw_transcripts_ready"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

print("YT_TRANSCRIPTS_DIR:", YT_TRANSCRIPTS_DIR)
print("EGO4D_TRANSCRIPTS_DIR:", EGO4D_TRANSCRIPTS_DIR)
print("YT_OUTCOMES_DIR:", YT_OUTCOMES_DIR)
print("EGO4D_OUTCOMES_DIR:", EGO4D_OUTCOMES_DIR)
print("OUTPUT_DIR:", OUTPUT_DIR)

YT_TRANSCRIPTS_DIR: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\raw\lai2023\Youtube\transcripts
EGO4D_TRANSCRIPTS_DIR: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\raw\lai2023\Ego4D\transcripts
YT_OUTCOMES_DIR: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\raw\lai2023\Youtube\vote_outcome_youtube_released
EGO4D_OUTCOMES_DIR: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\raw\lai2023\Ego4D\vote_outcome_ego4d
OUTPUT_DIR: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\processed\lai2023\onuw_transcripts_ready


In [4]:
from pathlib import Path
import re
from collections import defaultdict

EGO4D_MERGED_DIR = PROCESSED_DIR / "lai2023" / "ego4d_merged_transcripts"
EGO4D_MERGED_DIR.mkdir(parents=True, exist_ok=True)

EGO4D_CHUNK_RE = re.compile(
    r"^(?P<session>.+?)_(?P<game>Game\d+)_(?P<chunk>\d+)\.txt$",
    re.IGNORECASE
)

EGO4D_SINGLE_RE = re.compile(
    r"^(?P<session>.+?)_(?P<game>Game\d+)\.txt$",
    re.IGNORECASE
)

In [5]:
def merge_ego4d_transcripts(input_dir: Path, output_dir: Path):
    grouped = defaultdict(list)
    passthrough_files = []
    failed = []

    for path in sorted(input_dir.rglob("*.txt")):
        name = path.name

        m_chunk = EGO4D_CHUNK_RE.match(name)
        if m_chunk:
            session = m_chunk.group("session")
            game = m_chunk.group("game")
            chunk = int(m_chunk.group("chunk"))
            grouped[(session, game)].append((chunk, path))
            continue

        m_single = EGO4D_SINGLE_RE.match(name)
        if m_single:
            passthrough_files.append(path)
            continue

        failed.append((path, "Filename did not match expected Ego4D patterns"))

    merged_count = 0
    copied_count = 0

    # Merge chunked files
    for (session, game), chunk_items in grouped.items():
        chunk_items = sorted(chunk_items, key=lambda x: x[0])
        merged_texts = []

        for chunk_id, chunk_path in chunk_items:
            text = chunk_path.read_text(encoding="utf-8").strip()
            if text:
                merged_texts.append(text)

        merged_text = "\n".join(merged_texts).strip() + "\n"
        out_path = output_dir / f"{session}_{game}.txt"
        out_path.write_text(merged_text, encoding="utf-8")
        merged_count += 1

    # Copy already-single files
    for path in passthrough_files:
        out_path = output_dir / path.name
        out_path.write_text(path.read_text(encoding="utf-8"), encoding="utf-8")
        copied_count += 1

    return {
        "merged_count": merged_count,
        "copied_count": copied_count,
        "failed": failed,
    }
merge_result = merge_ego4d_transcripts(
    input_dir=EGO4D_TRANSCRIPTS_DIR,
    output_dir=EGO4D_MERGED_DIR,
)

print("Merged games:", merge_result["merged_count"])
print("Copied single-file games:", merge_result["copied_count"])
print("Failed:", len(merge_result["failed"]))

if merge_result["failed"]:
    for fp, err in merge_result["failed"][:10]:
        print("-", fp, "->", err)

Merged games: 48
Copied single-file games: 0
Failed: 0


In [6]:
TIMESTAMPED_LINE_RE = re.compile(
    r"^\s*(?P<speaker>.+?)\s*\((?P<timestamp>\d{1,2}:\d{2}(?::\d{2})?)\)\s*:\s*(?P<utterance>.*)\s*$"
)

PLAIN_LINE_RE = re.compile(
    r"^\s*(?P<speaker>[^:]+?)\s*:\s*(?P<utterance>.*)\s*$"
)

TRANSCRIPT_FILENAME_RE = re.compile(
    r"^(?P<session>.+?)_(?P<game>Game\d+)\.txt$",
    re.IGNORECASE
)

# Speaker tables live in src/utils/speakers.py, shared with the annotation
# analysis and the vote choice models.
import sys

if str(REPO_ROOT / "src") not in sys.path:
    sys.path.insert(0, str(REPO_ROOT / "src"))

from utils.speakers import (ANNOUNCER_LABEL, NON_PLAYER_SPEAKERS,
                            normalize_transcript_speaker)


def normalize_whitespace(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def normalize_speaker(speaker: str) -> str:
    """Roster spelling, with announcer-like speakers collapsed to one label."""
    return normalize_transcript_speaker(speaker)

In [7]:
from typing import Tuple, Dict
def canonicalize_session_name(name: str) -> str:
    """
    Normalize transcript/outcome names so they can be matched robustly.
    Example:
    'Flashback##ONE#NIGHT#ULTIMATE#WEREWOLF##Retro#9'
    -> 'flashback one night ultimate werewolf retro 9'
    """
    name = Path(name).stem  # remove .json if present
    name = name.replace("#", " ")
    name = re.sub(r"[_\-]+", " ", name)
    name = re.sub(r"\s+", " ", name)
    return name.strip().lower()


def parse_transcript_filename(path: Path) -> Tuple[str, str]:
    """
    Example:
    some_session_Game3.txt -> ("some_session", "Game3")
    """
    m = TRANSCRIPT_FILENAME_RE.match(path.name)
    if not m:
        raise ValueError(f"Transcript filename does not match expected pattern: {path.name}")
    return m.group("session"), m.group("game")


def build_outcome_lookup(outcomes_dir: Path) -> Dict[str, Path]:
    """
    Build a map from canonicalized session name -> actual outcome json path.
    """
    lookup = {}
    for outcome_path in outcomes_dir.glob("*.json"):
        key = canonicalize_session_name(outcome_path.stem)
        if key in lookup:
            print(f"Warning: duplicate canonical outcome key '{key}'")
            print(" -", lookup[key])
            print(" -", outcome_path)
        lookup[key] = outcome_path
    return lookup


def load_outcome_json_for_session(session_name: str, outcome_lookup: Dict[str, Path]) -> Tuple[Path, Dict]:
    key = canonicalize_session_name(session_name)

    if key not in outcome_lookup:
        raise FileNotFoundError(
            f"Outcome file not found for session '{session_name}' "
            f"(canonical key: '{key}')"
        )

    outcome_path = outcome_lookup[key]

    with outcome_path.open("r", encoding="utf-8") as f:
        outcome_data = json.load(f)

    return outcome_path, outcome_data


def get_game_metadata(outcome_data: Dict, game_key: str) -> Dict:
    if game_key not in outcome_data:
        raise KeyError(f"{game_key} not found in outcome JSON.")
    return outcome_data[game_key]

In [8]:
def parse_transcript_line(line: str) -> Optional[Dict[str, str]]:
    line = line.strip()
    if not line:
        return None

    m = TIMESTAMPED_LINE_RE.match(line)
    if m:
        return {
            "speaker": normalize_speaker(m.group("speaker")),
            "utterance": normalize_whitespace(m.group("utterance")),
        }

    m = PLAIN_LINE_RE.match(line)
    if m:
        return {
            "speaker": normalize_speaker(m.group("speaker")),
            "utterance": normalize_whitespace(m.group("utterance")),
        }

    return None


def process_transcript_with_metadata(
    transcript_path: Path,
    source_name: str,
    player_names: List[str],
    game_metadata: Dict,
) -> Dict:
    """
    Keep only:
    - official players from outcome JSON
    - announcer-like lines -> normalized as 'Announcer'
    """
    allowed_players = set(player_names)

    raw_text = transcript_path.read_text(encoding="utf-8")
    raw_lines = raw_text.splitlines()

    kept_turns = []
    dropped_lines = []
    unparsed_lines = []

    for raw_line in raw_lines:
        parsed = parse_transcript_line(raw_line)

        if parsed is None:
            if raw_line.strip():
                unparsed_lines.append(raw_line.strip())
            continue

        speaker = parsed["speaker"]

        if speaker == "Announcer" or speaker in allowed_players:
            kept_turns.append(parsed)
        else:
            dropped_lines.append(raw_line.strip())

    for idx, turn in enumerate(kept_turns, start=1):
        turn["turn_id"] = idx

    session_name, game_key = parse_transcript_filename(transcript_path)

    formatted_lines = [
        f"Source: {source_name}",
        f"Session: {session_name}",
        f"Game: {game_key}",
        f"Players: {', '.join(player_names)}",
        "",
        "Transcript:"
    ]

    for turn in kept_turns:
        formatted_lines.append(
            f"[{turn['turn_id']}] {turn['speaker']}: {turn['utterance']}"
        )

    formatted_transcript = "\n".join(formatted_lines)

    return {
        "players": player_names,
        "turns": kept_turns,
        "formatted_transcript": formatted_transcript,
        "game_metadata": game_metadata,
        "unparsed_lines": unparsed_lines,
        "dropped_lines": dropped_lines,
    }

In [9]:
def to_repo_relative(path: Path, repo_root: Path) -> str:
    """
    Convert an absolute path to a POSIX-style path relative to the repo root.
    """
    return path.resolve().relative_to(repo_root.resolve()).as_posix()
def save_processed_game(
    transcript_path: Path,
    output_dir: Path,
    source_name: str,
    session_name: str,
    game_key: str,
    processed: Dict,
) -> Dict[str, Path]:
    source_output_dir = output_dir / source_name
    source_output_dir.mkdir(parents=True, exist_ok=True)

    base_name = f"{session_name}_{game_key}"
    txt_path = source_output_dir / f"{base_name}.txt"

    txt_path.write_text(processed["formatted_transcript"], encoding="utf-8")

    return {"txt_path": txt_path}

In [10]:
def process_source_dataset(
    source_name: str,
    transcripts_dir: Path,
    outcomes_dir: Path,
    output_dir: Path,
) -> Tuple[List[Dict], List[Tuple[Path, str]]]:
    transcript_files = sorted(transcripts_dir.rglob("*.txt"))
    outcome_lookup = build_outcome_lookup(outcomes_dir)

    print(f"[{source_name}] Found {len(transcript_files)} transcript files.")
    print(f"[{source_name}] Found {len(outcome_lookup)} outcome json files.")

    saved = []
    failed = []

    for transcript_path in transcript_files:
        try:
            session_name, game_key = parse_transcript_filename(transcript_path)
            outcome_path, outcome_data = load_outcome_json_for_session(session_name, outcome_lookup)
            game_metadata = get_game_metadata(outcome_data, game_key)

            player_names = game_metadata["playerNames"]

            processed = process_transcript_with_metadata(
                transcript_path=transcript_path,
                source_name=source_name,
                player_names=player_names,
                game_metadata=game_metadata,
            )

            out_paths = save_processed_game(
                transcript_path=transcript_path,
                output_dir=output_dir,
                source_name=source_name,
                session_name=session_name,
                game_key=game_key,
                processed=processed,
            )

            for dropped_line in processed["dropped_lines"]:
                # Re-parse rather than splitting on ":", which would swallow the
                # timestamp into the name ("Matt (04").
                parsed_dropped = parse_transcript_line(dropped_line)
                dropped_speaker = (
                    parsed_dropped["speaker"] if parsed_dropped else dropped_line[:40]
                )
                dropped_speaker_counter[dropped_speaker] += 1

            saved.append({
                "source": source_name,
                "session_name": session_name,
                "game_key": game_key,
                "source_transcript_file": to_repo_relative(transcript_path, REPO_ROOT),
                "source_outcome_file": to_repo_relative(outcome_path, REPO_ROOT),
                "processed_txt_path": to_repo_relative(out_paths["txt_path"], REPO_ROOT),
                "num_players": len(player_names),
                "num_turns": len(processed["turns"]),
                "num_unparsed_lines": len(processed["unparsed_lines"]),
                "num_dropped_lines": len(processed["dropped_lines"]),
                "player_names": player_names,
                "voting_outcome": game_metadata.get("votingOutcome"),
                "start_roles": game_metadata.get("startRoles"),
                "end_roles": game_metadata.get("endRoles"),
                "warning": game_metadata.get("warning"),
            })

        except Exception as e:
            failed.append((transcript_path, str(e)))

    print(f"[{source_name}] Processed successfully: {len(saved)}")
    print(f"[{source_name}] Failed: {len(failed)}")

    return saved, failed

In [11]:
from collections import Counter

all_saved = []
all_failed = []

# Any speaker dropped at parse time is recorded, so a misspelled roster name can
# never again cost a player their turns without that showing up here.
dropped_speaker_counter = Counter()

yt_saved, yt_failed = process_source_dataset(
    source_name="Youtube",
    transcripts_dir=YT_TRANSCRIPTS_DIR,
    outcomes_dir=YT_OUTCOMES_DIR,
    output_dir=OUTPUT_DIR,
)
all_saved.extend(yt_saved)
all_failed.extend(yt_failed)

ego_saved, ego_failed = process_source_dataset(
    source_name="Ego4D",
    transcripts_dir=EGO4D_MERGED_DIR,
    outcomes_dir=EGO4D_OUTCOMES_DIR,
    output_dir=OUTPUT_DIR,
)
all_saved.extend(ego_saved)
all_failed.extend(ego_failed)

print("\nTOTAL PROCESSED:", len(all_saved))
print("TOTAL FAILED:", len(all_failed))

if all_failed:
    print("\nFirst failures:")
    for fp, err in all_failed[:10]:
        print(f"- {fp}: {err}")

print()
print("Speakers dropped at parse time (neither a roster player nor the announcer):")
if dropped_speaker_counter:
    for dropped_speaker, count in dropped_speaker_counter.most_common():
        expected = dropped_speaker.lower() in NON_PLAYER_SPEAKERS
        flag = "" if expected else "   <- UNEXPECTED: check for a misspelled roster name"
        print(f"  {dropped_speaker!r:24s} {count:3d} lines{flag}")
else:
    print("  none")

[Youtube] Found 151 transcript files.
[Youtube] Found 57 outcome json files.


[Youtube] Processed successfully: 151
[Youtube] Failed: 0
[Ego4D] Found 48 transcript files.
[Ego4D] Found 11 outcome json files.
[Ego4D] Processed successfully: 48
[Ego4D] Failed: 0

TOTAL PROCESSED: 199
TOTAL FAILED: 0

Speakers dropped at parse time (neither a roster player nor the announcer):
  'Matt'                     9 lines   <- UNEXPECTED: check for a misspelled roster name
  'Friend'                   4 lines
  'All'                      2 lines
  'New Speaker'              1 lines
  'Everyone'                 1 lines
  'Unknown'                  1 lines
  'Kaelan and Daniel'        1 lines


In [12]:
index_path = OUTPUT_DIR / "index.json"

# Guard against overwriting a good index with an empty one (e.g. if the raw data
# directory moves and no transcripts are found).
if not all_saved:
    raise RuntimeError(
        "No games were processed; refusing to overwrite the existing index. "
        f"Check that transcripts exist under {RAW_DIR}."
    )

index_path.write_text(
    json.dumps(all_saved, indent=2, ensure_ascii=False),
    encoding="utf-8"
)

print("Saved index to:", index_path)
print("Games in index:", len(all_saved))

Saved index to: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\data\processed\lai2023\onuw_transcripts_ready\index.json
Games in index: 199


In [13]:
import json
from pathlib import Path

index_path = OUTPUT_DIR / "index.json"

with index_path.open("r", encoding="utf-8") as f:
    index_data = json.load(f)

non_na_warnings = []

for row in index_data:
    warning = row.get("warning", "N/A")
    if warning is None:
        continue
    if str(warning).strip().lower() not in {"n/a", "na", ""}:
        non_na_warnings.append({
            "source": row.get("source"),
            "session_name": row.get("session_name"),
            "game_key": row.get("game_key"),
            "warning": warning,
        })

print(f"Found {len(non_na_warnings)} non-N/A warnings.\n")

for item in non_na_warnings:
    print(
        f"[{item['source']}] {item['session_name']} - {item['game_key']}: {item['warning']}"
    )

Found 22 non-N/A warnings.

[Ego4D] 0a6ef9dc-a2dc-452b-a907-d6fa2ed4cae0 - Game1: Jack is the moderator and so did not vote
[Ego4D] 0a6ef9dc-a2dc-452b-a907-d6fa2ed4cae0 - Game2: Erin is the moderator. No one voted anyone
[Ego4D] 0c2659db-7bd4-4b37-9b08-4e247befe382 - Game5: Doppelganger becomes Insomniac and Ashley incorrecly votes for self
[Ego4D] 0c2659db-7bd4-4b37-9b08-4e247befe382 - Game7: Doppelganger becomes Robber
[Ego4D] 0c2659db-7bd4-4b37-9b08-4e247befe382 - Game8: Doppelganger becomes Robber
[Ego4D] 1a0d3e22-a6d6-4723-91a0-52fc2c87d5ce - Game10: Jacob incorrectly looks at center card
[Ego4D] 324bccd2-adff-4e9a-8ff4-03491e144ac3 - Game1: Hunter incorrectly checks own card at end of night phase
[Ego4D] 3ba069be-60fa-47fc-bd7b-f85bf649a5bd - Game1: Daniels new card was not shown, it was from the center. player 1, 2 and 4 voted middle
[Ego4D] 3ba069be-60fa-47fc-bd7b-f85bf649a5bd - Game2: player 1, 2 and 3 voted middle
[Ego4D] 3ba069be-60fa-47fc-bd7b-f85bf649a5bd - Game3: All play

In [14]:
import json
from collections import Counter

index_path = OUTPUT_DIR / "index.json"

with index_path.open("r", encoding="utf-8") as f:
    index_data = json.load(f)

start_role_counter = Counter()
end_role_counter = Counter()

for row in index_data:
    start_role_counter.update(row.get("start_roles", []))
    end_role_counter.update(row.get("end_roles", []))

all_roles = sorted(set(start_role_counter) | set(end_role_counter))

print("Unique roles in dataset:")
for role in all_roles:
    print(f"- {role}")

print("\nStart role counts:")
for role, count in start_role_counter.most_common():
    print(f"{role}: {count}")

print("\nEnd role counts:")
for role, count in end_role_counter.most_common():
    print(f"{role}: {count}")

Unique roles in dataset:
- Center Card
- Center card
- Doppelganger
- Doppleganger
- Drunk
- Hunter
- Insomniac
- Mason
- Minion
- Minions of Mordred
- Moderator
- NA
- Revealer
- Robber
- Seer
- Tanner
- Troublemaker
- Villager
- Werewolf

Start role counts:
Werewolf: 221
Robber: 118
Troublemaker: 115
Seer: 95
Tanner: 84
Villager: 70
Insomniac: 48
Drunk: 45
NA: 38
Minion: 27
Hunter: 20
Mason: 12
Moderator: 3
Doppelganger: 3
Doppleganger: 3
Minions of Mordred: 2

End role counts:
Werewolf: 229
Robber: 120
Troublemaker: 120
Seer: 99
Tanner: 84
Villager: 73
Insomniac: 49
NA: 38
Minion: 29
Hunter: 22
Drunk: 13
Mason: 11
Center card: 5
Moderator: 3
Doppelganger: 3
Doppleganger: 2
Minions of Mordred: 2
Revealer: 1
Center Card: 1


## Exclude invalid games

This is the **only** cell that writes `index_cleaned.json` and `index_excluded.json`.
It does four things, in order:

1. normalizes role spellings (`Center card` -> `Center Card`, `Doppleganger` -> `Doppelganger`);
2. normalizes `warning` from the raw string field into a **list of warning strings**
   (`"N/A"` becomes an empty list), so downstream code can always treat it as a list;
3. excludes games that cannot support any analysis:
   - `different_game`: the `warning` says the group played a different game (Avalon, not ONUW);
   - `na_role`: `startRoles`, `endRoles` and `votingOutcome` are all the literal string
     `"NA"`, i.e. the release carries no ground truth for that game;
4. appends a derived warning, `No 'Werewolf' role in end roles`, to kept games where no
   player holds the Werewolf card at voting time. Those games are **kept** — both wolf
   cards ending in the center is legal ONUW — but they need flagging, because "the village
   voted correctly" means something different when no correct wolf-vote was available.

All steps are idempotent: re-running the cell reproduces the same files.

In [15]:
import json
from pathlib import Path
from collections import Counter

index_path = OUTPUT_DIR / "index.json"
clean_index_path = OUTPUT_DIR / "index_cleaned.json"
excluded_index_path = OUTPUT_DIR / "index_excluded.json"

with index_path.open("r", encoding="utf-8") as f:
    index_data = json.load(f)

ROLE_NORMALIZATION = {
    "Center card": "Center Card",
    "Doppleganger": "Doppelganger",
}

# Values that mean "no warning recorded" in the released outcome JSONs.
EMPTY_WARNING_VALUES = {"n/a", "na", "none", ""}

NO_WEREWOLF_WARNING = "No 'Werewolf' role in end roles"


def normalize_role(role):
    if role is None:
        return role
    role = str(role).strip()
    return ROLE_NORMALIZATION.get(role, role)


def normalize_warning(warning) -> List[str]:
    """
    Convert the raw `warning` field into a list of warning strings.

    The released JSONs store a single string, using 'N/A' for "no warning".
    Accepting a list as input as well makes this idempotent, so the cell can be
    re-run on an already-cleaned index without corrupting it.
    """
    if warning is None:
        return []

    items = warning if isinstance(warning, list) else [warning]

    normalized = []
    for item in items:
        text = str(item).strip()
        if text.lower() in EMPTY_WARNING_VALUES:
            continue
        normalized.append(text)

    return normalized


def has_na_role(row) -> bool:
    all_roles = [*row.get("start_roles", []), *row.get("end_roles", [])]
    return any(str(r).strip().upper() == "NA" for r in all_roles)


def exclusion_reason(row) -> Optional[str]:
    """
    Reason to drop a game, or None to keep it. `different_game` takes priority:
    those games also have NA roles, but the more specific reason is more informative.
    """
    warnings = row.get("warning", [])

    if any("different game" in w.lower() for w in warnings):
        return "different_game"
    if has_na_role(row):
        return "na_role"
    return None


clean_rows = []
excluded_rows = []
normalization_stats = Counter()

for row in index_data:
    new_row = dict(row)

    for role_field in ("start_roles", "end_roles"):
        old_roles = row.get(role_field, [])
        new_roles = []
        for r in old_roles:
            nr = normalize_role(r)
            if nr != r:
                normalization_stats[f"{role_field}: {r} -> {nr}"] += 1
            new_roles.append(nr)
        new_row[role_field] = new_roles

    new_row["warning"] = normalize_warning(row.get("warning"))

    reason = exclusion_reason(new_row)
    if reason is not None:
        new_row["exclusion_reason"] = reason
        excluded_rows.append(new_row)
    else:
        clean_rows.append(new_row)

# Derived flag on kept games only: excluded games have no usable roles, so flagging
# them for a missing Werewolf would be meaningless.
n_no_werewolf_end = 0

for row in clean_rows:
    if "Werewolf" not in row["end_roles"]:
        n_no_werewolf_end += 1
        if NO_WEREWOLF_WARNING not in row["warning"]:
            row["warning"].append(NO_WEREWOLF_WARNING)

with clean_index_path.open("w", encoding="utf-8") as f:
    json.dump(clean_rows, f, indent=2, ensure_ascii=False)

with excluded_index_path.open("w", encoding="utf-8") as f:
    json.dump(excluded_rows, f, indent=2, ensure_ascii=False)

print(f"Original rows: {len(index_data)}")
print(f"Cleaned rows kept: {len(clean_rows)}")
print(f"Excluded rows: {len(excluded_rows)}")

print("\nRole normalizations applied:")
for k, v in normalization_stats.items():
    print(f"- {k}: {v}")

print(f"\nKept games flagged '{NO_WEREWOLF_WARNING}': {n_no_werewolf_end}")

print("\nExclusions by reason:")
for reason, count in Counter(r["exclusion_reason"] for r in excluded_rows).most_common():
    print(f"- {reason}: {count}")

if excluded_rows:
    print("\nExcluded games:")
    for row in excluded_rows:
        print(
            f"- [{row['source']}] {row['session_name']} - {row['game_key']} "
            f"| reason={row['exclusion_reason']} | warning={row['warning']}"
        )

print("\nSaved:", clean_index_path)
print("Saved:", excluded_index_path)

Original rows: 199
Cleaned rows kept: 191
Excluded rows: 8

Role normalizations applied:
- end_roles: Center card -> Center Card: 5
- start_roles: Doppleganger -> Doppelganger: 3
- end_roles: Doppleganger -> Doppelganger: 2

Kept games flagged 'No 'Werewolf' role in end roles': 26

Exclusions by reason:
- na_role: 7
- different_game: 1

Excluded games:
- [Ego4D] 0d495a56-676b-4bb0-bb7d-9abf88fb7beb - Game1 | reason=na_role | warning=[]
- [Ego4D] 0d495a56-676b-4bb0-bb7d-9abf88fb7beb - Game2 | reason=na_role | warning=[]
- [Ego4D] 0d495a56-676b-4bb0-bb7d-9abf88fb7beb - Game3 | reason=na_role | warning=[]
- [Ego4D] 0d495a56-676b-4bb0-bb7d-9abf88fb7beb - Game4 | reason=na_role | warning=[]
- [Ego4D] 1281ef80-2747-4bd1-8061-373409c879b5 - Game5 | reason=na_role | warning=[]
- [Ego4D] 1281ef80-2747-4bd1-8061-373409c879b5 - Game6 | reason=na_role | warning=[]
- [Ego4D] 1281ef80-2747-4bd1-8061-373409c879b5 - Game7 | reason=na_role | warning=[]
- [Ego4D] 64076f60-f3fc-4ee6-93e1-3489d8ec7b12 - G

In [16]:
# Read-only verification of what the previous cell wrote.
from collections import Counter

with clean_index_path.open("r", encoding="utf-8") as f:
    cleaned_index = json.load(f)

start_role_counter = Counter()
end_role_counter = Counter()

for row in cleaned_index:
    start_role_counter.update(row["start_roles"])
    end_role_counter.update(row["end_roles"])

all_roles = sorted(set(start_role_counter) | set(end_role_counter))

print("Unique roles after cleaning:")
for role in all_roles:
    print(f"- {role} (start: {start_role_counter[role]}, end: {end_role_counter[role]})")

# Werewolf presence. The end-role count is the one that matters for outcome analysis.
no_ww_start = sum(1 for row in cleaned_index if "Werewolf" not in row["start_roles"])
no_ww_end = sum(1 for row in cleaned_index if "Werewolf" not in row["end_roles"])
no_ww_both = sum(
    1 for row in cleaned_index
    if "Werewolf" not in row["start_roles"] and "Werewolf" not in row["end_roles"]
)

print("\nGames without a Werewolf:")
print(f"- no Werewolf in start roles (both wolf cards dealt to center): {no_ww_start}")
print(f"- no Werewolf in end roles (no wolf at voting time): {no_ww_end}")
print(f"- no Werewolf in start and end roles: {no_ww_both}")

# Schema checks: `warning` must be a list everywhere, and the derived flag must agree
# with the end roles.
non_list_warnings = [row for row in cleaned_index if not isinstance(row["warning"], list)]
flagged = [row for row in cleaned_index if NO_WEREWOLF_WARNING in row["warning"]]
na_roles_left = sum(
    1 for row in cleaned_index if "NA" in row["start_roles"] + row["end_roles"]
)

print("\nChecks:")
print(f"- rows where `warning` is not a list: {len(non_list_warnings)} (expected 0)")
print(f"- rows carrying '{NO_WEREWOLF_WARNING}': {len(flagged)} (expected {no_ww_end})")
print(f"- rows with 'NA' left in a role list: {na_roles_left} (expected 0)")

print("\nWarning inventory (kept games):")
warning_counter = Counter(w for row in cleaned_index for w in row["warning"])
for warning, count in warning_counter.most_common():
    print(f"- {count:3d}  {warning}")

n_with_warning = sum(1 for row in cleaned_index if row["warning"])
print(f"\nGames with at least one warning: {n_with_warning} / {len(cleaned_index)}")

Unique roles after cleaning:
- Center Card (start: 0, end: 6)
- Doppelganger (start: 6, end: 5)
- Drunk (start: 45, end: 13)
- Hunter (start: 20, end: 22)
- Insomniac (start: 48, end: 49)
- Mason (start: 12, end: 11)
- Minion (start: 27, end: 29)
- Moderator (start: 3, end: 3)
- Revealer (start: 0, end: 1)
- Robber (start: 118, end: 120)
- Seer (start: 95, end: 99)
- Tanner (start: 84, end: 84)
- Troublemaker (start: 115, end: 120)
- Villager (start: 70, end: 73)
- Werewolf (start: 221, end: 229)

Games without a Werewolf:
- no Werewolf in start roles (both wolf cards dealt to center): 28
- no Werewolf in end roles (no wolf at voting time): 26
- no Werewolf in start and end roles: 25

Checks:
- rows where `warning` is not a list: 0 (expected 0)
- rows carrying 'No 'Werewolf' role in end roles': 26 (expected 26)
- rows with 'NA' left in a role list: 0 (expected 0)

Warning inventory (kept games):
-  26  No 'Werewolf' role in end roles
-   4  All players voted middle
-   2  Doppelganger 

Player number distribution

In [17]:
import json
from collections import Counter

index_path = OUTPUT_DIR / "index_cleaned.json"  # or index.json if you prefer

with index_path.open("r", encoding="utf-8") as f:
    index_data = json.load(f)

num_players_counter = Counter()

for row in index_data:
    num_players = len(row.get("player_names", []))
    num_players_counter[num_players] += 1

print("Player count distribution across games:\n")
for n_players in sorted(num_players_counter):
    print(f"{n_players} players: {num_players_counter[n_players]} games")

Player count distribution across games:

3 players: 8 games
4 players: 100 games
5 players: 58 games
6 players: 25 games


## Length and composition statistics

Descriptive statistics for the cleaned corpus (`index_cleaned.json`), computed from the
processed transcripts under `onuw_transcripts_ready/`.

Two things to keep in mind when reporting these numbers:

- `num_turns` in the index counts **all** kept lines, including `Announcer` lines. The
  tables below separate announcer turns from player turns.
- These counts come from the transcripts parsed here. Lai et al.'s strategy-annotation
  files contain a slightly different (usually smaller) number of turns for the same games;
  see the matching sanity check in `persuasion_strategy_distribution.ipynb`.

Standard deviations are sample standard deviations (`ddof=1`).

In [18]:
import pandas as pd

STATS_DIR = REPO_ROOT / "analysis" / "lai2023_corpus" / "tables"
STATS_DIR.mkdir(parents=True, exist_ok=True)

PROCESSED_LINE_RE = re.compile(
    r"^\[(?P<turn_id>\d+)\]\s+(?P<speaker>.+?):\s?(?P<utterance>.*)$"
)

ANNOUNCER_SPEAKER = "Announcer"


def read_processed_turns(processed_txt_path: str) -> Tuple[List[Dict], int]:
    """
    Re-read a processed transcript and return its turns plus the number of
    transcript-body lines that did not match the expected output format.
    """
    path = REPO_ROOT / processed_txt_path
    text = path.read_text(encoding="utf-8")

    # Everything after the "Transcript:" header line is the dialogue body.
    body = text.split("Transcript:", 1)[1]

    turns = []
    malformed = 0

    for line in body.splitlines():
        if not line.strip():
            continue

        m = PROCESSED_LINE_RE.match(line.strip())
        if m is None:
            malformed += 1
            continue

        utterance = m.group("utterance").strip()
        turns.append({
            "speaker": m.group("speaker"),
            "utterance": utterance,
            "n_words": len(utterance.split()),
            "n_chars": len(utterance),
        })

    return turns, malformed


def build_game_stats(index_rows: List[Dict]) -> Tuple[pd.DataFrame, pd.DataFrame]:
    """
    Build a per-game and a per-turn table of length statistics.
    """
    game_rows = []
    turn_rows = []

    for row in index_rows:
        turns, malformed = read_processed_turns(row["processed_txt_path"])

        announcer_turns = [t for t in turns if t["speaker"] == ANNOUNCER_SPEAKER]
        player_turns = [t for t in turns if t["speaker"] != ANNOUNCER_SPEAKER]

        game_id = f"{row['source']} / {row['session_name']} / {row['game_key']}"

        for turn in turns:
            turn_rows.append({
                "game_id": game_id,
                "source": row["source"],
                "speaker": turn["speaker"],
                "is_announcer": turn["speaker"] == ANNOUNCER_SPEAKER,
                "n_words": turn["n_words"],
                "n_chars": turn["n_chars"],
            })

        game_rows.append({
            "game_id": game_id,
            "source": row["source"],
            "session_name": row["session_name"],
            "game_key": row["game_key"],
            "num_players": len(row["player_names"]),
            "num_turns_index": row["num_turns"],
            "n_turns": len(turns),
            "n_announcer_turns": len(announcer_turns),
            "n_player_turns": len(player_turns),
            "n_words": sum(t["n_words"] for t in turns),
            "n_words_player_turns": sum(t["n_words"] for t in player_turns),
            "n_chars": sum(t["n_chars"] for t in turns),
            "n_malformed_lines": malformed,
            "num_dropped_lines": row["num_dropped_lines"],
            "num_unparsed_lines": row["num_unparsed_lines"],
            "has_werewolf_start": "Werewolf" in row["start_roles"],
            "has_werewolf_end": "Werewolf" in row["end_roles"],
        })

    return pd.DataFrame(game_rows), pd.DataFrame(turn_rows)


with clean_index_path.open("r", encoding="utf-8") as f:
    cleaned_index = json.load(f)

game_stats_df, turn_stats_df = build_game_stats(cleaned_index)

game_stats_df.to_csv(STATS_DIR / "game_length_composition.csv", index=False)

print("Games:", len(game_stats_df))
print("Turns:", len(turn_stats_df))
print("Malformed transcript-body lines:", int(game_stats_df["n_malformed_lines"].sum()))
print()
print("Consistency check against index `num_turns`:")
print(
    "  games where n_turns == num_turns_index:",
    int((game_stats_df["n_turns"] == game_stats_df["num_turns_index"]).sum()),
    "/",
    len(game_stats_df),
)
print()
print("Saved:", STATS_DIR / "game_length_composition.csv")

display(game_stats_df.head())

Games: 191
Turns: 25384
Malformed transcript-body lines: 0

Consistency check against index `num_turns`:
  games where n_turns == num_turns_index: 191 / 191

Saved: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\game_length_composition.csv


,game_id,source,session_name,game_key,num_players,num_turns_index,n_turns,n_announcer_turns,n_player_turns,n_words,n_words_player_turns,n_chars,n_malformed_lines,num_dropped_lines,num_unparsed_lines,has_werewolf_start,has_werewolf_end
0,Youtube / Flashback##ONE#NIGHT#ULTIMATE#WEREWO...,Youtube,Flashback##ONE#NIGHT#ULTIMATE#WEREWOLF##Retro#9,Game1,5,112,112,5,107,1119,993,5864,0,0,0,True,True
1,Youtube / Flashback##ONE#NIGHT#ULTIMATE#WEREWO...,Youtube,Flashback##ONE#NIGHT#ULTIMATE#WEREWOLF##Retro#9,Game2,5,157,157,10,147,1206,1124,6206,0,0,0,True,True
2,Youtube / Flashback##ONE#NIGHT#ULTIMATE#WEREWO...,Youtube,Flashback##ONE#NIGHT#ULTIMATE#WEREWOLF##Retro#9,Game3,5,79,79,9,70,784,680,4136,0,0,0,True,True
3,Youtube / Flashback##ONE#NIGHT#ULTIMATE#WEREWO...,Youtube,Flashback##ONE#NIGHT#ULTIMATE#WEREWOLF##Retro#9,Game4,5,148,148,9,139,1172,1067,5946,0,0,0,False,False
4,Youtube / Flashback##ONE#NIGHT#ULTIMATE#WEREWO...,Youtube,Flashback##ONE#NIGHT#ULTIMATE#WEREWOLF##Retro#9,Game5,5,85,85,5,80,967,846,5104,0,0,0,True,True


### Length descriptives

Game-level and turn-level length, overall and split by source. Utterance length is
strongly right-skewed (spoken multi-party dialogue), so the median is the number to
report rather than the mean.

In [19]:
def describe_series(name: str, group: str, series: pd.Series) -> Dict:
    """Standard descriptive row for a numeric series."""
    return {
        "measure": name,
        "group": group,
        "n": int(series.size),
        "total": int(series.sum()),
        "mean": round(float(series.mean()), 2),
        "sd": round(float(series.std(ddof=1)), 2),
        "min": int(series.min()),
        "q25": float(series.quantile(0.25)),
        "median": float(series.median()),
        "q75": float(series.quantile(0.75)),
        "p90": float(series.quantile(0.90)),
        "max": int(series.max()),
    }


MEASURES = [
    ("turns per game", game_stats_df, "n_turns"),
    ("player turns per game", game_stats_df, "n_player_turns"),
    ("words per game", game_stats_df, "n_words"),
    ("characters per game", game_stats_df, "n_chars"),
    ("players per game", game_stats_df, "num_players"),
    ("words per turn", turn_stats_df, "n_words"),
]

summary_rows = []

for measure_name, frame, column in MEASURES:
    summary_rows.append(describe_series(measure_name, "all", frame[column]))
    for source in sorted(frame["source"].unique()):
        summary_rows.append(
            describe_series(
                measure_name,
                source,
                frame.loc[frame["source"] == source, column],
            )
        )

length_summary_df = pd.DataFrame(summary_rows)
length_summary_df.to_csv(STATS_DIR / "length_summary.csv", index=False)

print("Length descriptives (cleaned corpus)")
print("-" * 80)
display(length_summary_df)

print("\nSaved:", STATS_DIR / "length_summary.csv")

Length descriptives (cleaned corpus)
--------------------------------------------------------------------------------


,measure,group,n,total,mean,sd,min,q25,median,q75,p90,max
0,turns per game,all,191,25384,132.90,57.65,20,97.00,130.0,158.50,199.0,403
1,turns per game,Ego4D,40,5325,133.12,70.66,20,66.00,131.5,187.25,220.6,284
2,turns per game,Youtube,151,20059,132.84,53.96,23,102.00,129.0,156.50,189.0,403
3,player turns per game,all,191,24100,126.18,56.27,18,91.00,123.0,151.00,189.0,384
4,player turns per game,Ego4D,40,5255,131.38,70.22,20,64.50,131.0,186.50,219.4,281
5,player turns per game,Youtube,151,18845,124.80,52.14,18,97.00,120.0,146.50,180.0,384
6,words per game,all,191,189741,993.41,383.82,108,800.50,1014.0,1173.50,1361.0,3173
7,words per game,Ego4D,40,37576,939.40,510.54,108,421.00,1025.0,1357.00,1551.8,1909
8,words per game,Youtube,151,152165,1007.72,343.30,281,832.00,1014.0,1166.00,1283.0,3173
9,characters per game,all,191,979076,5126.05,1971.10,570,4116.50,5246.0,6026.00,7072.0,16063



Saved: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\length_summary.csv


### Corpus composition

Table size by source, speaker inventory, announcer share, parsing losses, and
Werewolf presence.

`has_werewolf_end` is the relevant flag for outcome analysis: voting is scored against
end roles, so a game with no Werewolf at voting time offers no correct wolf-vote,
regardless of what happened during the night.

In [20]:
# Table size by source.
player_dist_df = (
    game_stats_df
    .pivot_table(
        index="num_players",
        columns="source",
        values="game_id",
        aggfunc="count",
        fill_value=0,
    )
)
player_dist_df["all"] = player_dist_df.sum(axis=1)
player_dist_df["percent_of_games"] = (
    100 * player_dist_df["all"] / player_dist_df["all"].sum()
).round(1)

player_dist_df.to_csv(STATS_DIR / "player_count_distribution.csv")

print("Games by table size:")
display(player_dist_df)

# Speaker inventory and sessions.
speaker_counter = Counter()
for row in cleaned_index:
    speaker_counter.update(row["player_names"])

sessions = {(row["source"], row["session_name"]) for row in cleaned_index}

print("\nSessions:", len(sessions))
for source in sorted({s for s, _ in sessions}):
    print(f"  {source}: {len({n for s, n in sessions if s == source})}")

print("\nPlayer slots (sum over games):", sum(len(r["player_names"]) for r in cleaned_index))
print("Unique player names:", len(speaker_counter))
print("Most frequent players:", speaker_counter.most_common(5))

# Announcer share and parsing losses.
n_turns = int(game_stats_df["n_turns"].sum())
n_announcer = int(game_stats_df["n_announcer_turns"].sum())

print("\nAnnouncer turns:", n_announcer, f"({n_announcer / n_turns:.2%} of all turns)")
print("Player turns:", n_turns - n_announcer)
print(
    "Lines dropped at parse time (non-player, non-announcer speakers):",
    int(game_stats_df["num_dropped_lines"].sum()),
)
print("Lines that could not be parsed at all:", int(game_stats_df["num_unparsed_lines"].sum()))

# Werewolf presence.
no_ww_start = int((~game_stats_df["has_werewolf_start"]).sum())
no_ww_end = int((~game_stats_df["has_werewolf_end"]).sum())
no_ww_both = int(
    (~game_stats_df["has_werewolf_start"] & ~game_stats_df["has_werewolf_end"]).sum()
)

print("\nWerewolf presence:")
print(f"  no Werewolf in start roles: {no_ww_start} ({no_ww_start / len(game_stats_df):.1%})")
print(f"  no Werewolf in end roles:   {no_ww_end} ({no_ww_end / len(game_stats_df):.1%})")
print(f"  no Werewolf in start and end roles: {no_ww_both}")

no_ww_by_source = (
    game_stats_df
    .assign(no_werewolf_end=~game_stats_df["has_werewolf_end"])
    .groupby("source")
    .agg(n_games=("game_id", "count"), n_no_werewolf_end=("no_werewolf_end", "sum"))
)
display(no_ww_by_source)

print("\nSaved:", STATS_DIR / "player_count_distribution.csv")

Games by table size:


source,Ego4D,Youtube,all,percent_of_games
num_players,,,,
3,0,8,8,4.2
4,28,72,100,52.4
5,3,55,58,30.4
6,9,16,25,13.1



Sessions: 66
  Ego4D: 9
  Youtube: 57

Player slots (sum over games): 864
Unique player names: 60
Most frequent players: [('Mitchell', 132), ('Justin', 120), ('Mike', 86), ('Paul', 74), ('James', 65)]

Announcer turns: 1284 (5.06% of all turns)
Player turns: 24100
Lines dropped at parse time (non-player, non-announcer speakers): 17
Lines that could not be parsed at all: 0

Werewolf presence:
  no Werewolf in start roles: 28 (14.7%)
  no Werewolf in end roles:   26 (13.6%)
  no Werewolf in start and end roles: 25


,n_games,n_no_werewolf_end
source,,
Ego4D,40,8
Youtube,151,18



Saved: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\player_count_distribution.csv


## Role distribution and night-phase role change

Role frequencies over the 864 player slots of the cleaned corpus, for the role dealt
before the night phase (`start_roles`) and the role held at voting time (`end_roles`),
plus the full start -> end transition matrix.

The unit is a **player slot** (one player in one game), not a person: the same players
recur across games.

Three roles are not standard base-game ONUW roles and are reported for completeness:
`Moderator` (3 slots, an Ego4D participant who ran the game instead of playing),
`Center Card` (an end role only, 6 slots, where the player's card ended up in the
centre) and `Revealer` (1 slot, from an expansion).

In [21]:
slot_rows = []

for row in cleaned_index:
    game_id = f"{row['source']} / {row['session_name']} / {row['game_key']}"

    for slot_index, (start_role, end_role) in enumerate(
        zip(row["start_roles"], row["end_roles"])
    ):
        slot_rows.append({
            "game_id": game_id,
            "source": row["source"],
            "slot_index": slot_index,
            "player_name": row["player_names"][slot_index],
            "start_role": start_role,
            "end_role": end_role,
            "role_changed": start_role != end_role,
        })

slot_df = pd.DataFrame(slot_rows)
slot_df.to_csv(STATS_DIR / "player_role_slots.csv", index=False)

n_slots = len(slot_df)
n_games_total = slot_df["game_id"].nunique()
roles = sorted(set(slot_df["start_role"]) | set(slot_df["end_role"]))

role_distribution_df = pd.DataFrame([
    {
        "role": role,
        "n_start": int((slot_df["start_role"] == role).sum()),
        "percent_start": 100 * (slot_df["start_role"] == role).mean(),
        "n_end": int((slot_df["end_role"] == role).sum()),
        "percent_end": 100 * (slot_df["end_role"] == role).mean(),
        "delta_n": int((slot_df["end_role"] == role).sum() - (slot_df["start_role"] == role).sum()),
        "delta_percentage_points": 100 * (
            (slot_df["end_role"] == role).mean() - (slot_df["start_role"] == role).mean()
        ),
        "n_games_start": int(slot_df.loc[slot_df["start_role"] == role, "game_id"].nunique()),
        "n_games_end": int(slot_df.loc[slot_df["end_role"] == role, "game_id"].nunique()),
    }
    for role in roles
]).sort_values("n_start", ascending=False).reset_index(drop=True)

role_distribution_df.to_csv(STATS_DIR / "role_distribution.csv", index=False)

# Full transition matrix, rows = start role, cols = end role.
role_transition_df = (
    pd.crosstab(slot_df["start_role"], slot_df["end_role"])
    .reindex(index=roles, columns=roles, fill_value=0)
)
role_transition_df.to_csv(STATS_DIR / "role_transition_matrix.csv")

n_changed = int(slot_df["role_changed"].sum())
games_with_change = int(slot_df.groupby("game_id")["role_changed"].any().sum())

print(f"Player slots: {n_slots} across {n_games_total} games")
print(f"Slots whose role changed during the night: {n_changed} ({100 * n_changed / n_slots:.1f}%)")
print(f"Games with at least one role change: {games_with_change} / {n_games_total}")

print("\nRole distribution (start vs end):")
display(role_distribution_df)

# Rules check: the Drunk always swaps its card with the centre and never keeps the role.
n_drunk_kept = int(role_transition_df.loc["Drunk", "Drunk"])
print(f"\nStart-Drunk slots that are still Drunk at voting time: {n_drunk_kept} "
      f"(expected 0 — the Drunk always swaps with a centre card)")

print("\nMost frequent role changes:")
transitions = [
    {"start_role": s, "end_role": e, "n": int(role_transition_df.loc[s, e])}
    for s in roles for e in roles
    if s != e and role_transition_df.loc[s, e] > 0
]
display(
    pd.DataFrame(transitions)
    .sort_values("n", ascending=False)
    .head(12)
    .reset_index(drop=True)
)

print("\nSaved:")
for filename in ["player_role_slots.csv", "role_distribution.csv", "role_transition_matrix.csv"]:
    print("-", STATS_DIR / filename)

Player slots: 864 across 191 games
Slots whose role changed during the night: 383 (44.3%)
Games with at least one role change: 156 / 191

Role distribution (start vs end):


,role,n_start,percent_start,n_end,percent_end,delta_n,delta_percentage_points,n_games_start,n_games_end
0,Werewolf,221,25.578704,229,26.504630,8,0.925926,163,165
1,Robber,118,13.657407,120,13.888889,2,0.231481,118,120
2,Troublemaker,115,13.310185,120,13.888889,5,0.578704,115,120
3,Seer,95,10.995370,99,11.458333,4,0.462963,95,99
4,Tanner,84,9.722222,84,9.722222,0,0.000000,84,84
5,Villager,70,8.101852,73,8.449074,3,0.347222,48,51
6,Insomniac,48,5.555556,49,5.671296,1,0.115741,48,49
7,Drunk,45,5.208333,13,1.504630,-32,-3.703704,45,13
8,Minion,27,3.125000,29,3.356481,2,0.231481,27,29
9,Hunter,20,2.314815,22,2.546296,2,0.231481,20,22



Start-Drunk slots that are still Drunk at voting time: 0 (expected 0 — the Drunk always swaps with a centre card)

Most frequent role changes:


,start_role,end_role,n
0,Robber,Werewolf,30
1,Werewolf,Robber,29
2,Robber,Seer,18
3,Troublemaker,Robber,18
4,Robber,Tanner,17
5,Seer,Robber,14
6,Robber,Villager,12
7,Villager,Robber,12
8,Tanner,Werewolf,12
9,Werewolf,Tanner,10



Saved:
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\player_role_slots.csv
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\role_distribution.csv
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\role_transition_matrix.csv


### Do the start and end role distributions differ?

The start and end role counts are *paired*: every player slot contributes one start role
and one end role, so an unpaired chi-square test of the two columns would be wrong. The
question is whether the **marginal** distribution shifts between the deal and the vote.

Three tests, from global to per-role:

1. **Stuart-Maxwell** test of marginal homogeneity on the paired transition matrix
   (the multi-category generalisation of McNemar). Categories with no discordant pairs
   are dropped, since they contribute nothing and make the covariance matrix singular.
2. **Exact McNemar** per role, on the counts of slots that lost the role (`b`) versus
   gained it (`c`), with **Holm** correction across roles.
3. **Game-clustered bootstrap** CI on each role's change in share, following the same
   convention as the role x strategy contrasts in
   `persuasion_strategy_distribution.ipynb`: resampling unit = game, 5,000 draws, seeded.

The first two treat player slots as independent, which they are not — a single Robber
swap changes two slots in the same game at once. The bootstrap is the one that respects
the clustering, so where they disagree, prefer the bootstrap for the direction and the
Holm-corrected McNemar for the claim of significance.

In [22]:
import numpy as np
from scipy import stats

ROLE_SEED = 20260729
ROLE_N_BOOT = 5_000
ROLE_CI = (2.5, 97.5)

transition = role_transition_df.to_numpy(dtype=float)

# --- 1. Stuart-Maxwell test of marginal homogeneity --------------------------
discordant_total = [
    (transition[i].sum() - transition[i, i]) + (transition[:, i].sum() - transition[i, i])
    for i in range(len(roles))
]
kept = [i for i, total in enumerate(discordant_total) if total > 0]
dropped_roles = [roles[i] for i in range(len(roles)) if i not in kept]

sub = transition[np.ix_(kept, kept)]
k = len(kept)

# Marginal differences, one category dropped for identifiability.
d = (sub.sum(axis=1) - sub.sum(axis=0))[:-1]

cov = np.zeros((k - 1, k - 1))
for i in range(k - 1):
    cov[i, i] = sub[i, :].sum() + sub[:, i].sum() - 2 * sub[i, i]
    for j in range(k - 1):
        if i != j:
            cov[i, j] = -(sub[i, j] + sub[j, i])

sm_chi2 = float(d @ np.linalg.pinv(cov) @ d)
sm_dof = int(np.linalg.matrix_rank(cov))
sm_p = float(stats.chi2.sf(sm_chi2, sm_dof))

print("1. STUART-MAXWELL TEST OF MARGINAL HOMOGENEITY")
print("-" * 78)
print(f"categories tested: {k} (dropped, no discordant pairs: {dropped_roles or 'none'})")
print(f"chi2 = {sm_chi2:.2f}, df = {sm_dof}, p = {sm_p:.4f}")
print(
    "-> the start and end role distributions are "
    f"{'NOT ' if sm_p >= 0.05 else ''}distinguishable at alpha = 0.05"
)

# --- 2. Exact McNemar per role, Holm-corrected -------------------------------
mcnemar_rows = []

for role in roles:
    lost = int(((slot_df["start_role"] == role) & (slot_df["end_role"] != role)).sum())
    gained = int(((slot_df["start_role"] != role) & (slot_df["end_role"] == role)).sum())
    n_discordant = lost + gained

    p_raw = (
        float(stats.binomtest(min(lost, gained), n_discordant, 0.5).pvalue)
        if n_discordant > 0 else 1.0
    )

    mcnemar_rows.append({
        "role": role,
        "n_start": int((slot_df["start_role"] == role).sum()),
        "n_end": int((slot_df["end_role"] == role).sum()),
        "n_lost": lost,
        "n_gained": gained,
        "p_mcnemar_raw": p_raw,
    })

role_test_df = pd.DataFrame(mcnemar_rows).sort_values("p_mcnemar_raw").reset_index(drop=True)

# Holm step-down correction.
n_tests = len(role_test_df)
holm_adjusted = []
running_max = 0.0
for i, p_raw in enumerate(role_test_df["p_mcnemar_raw"]):
    running_max = max(running_max, min(1.0, (n_tests - i) * p_raw))
    holm_adjusted.append(running_max)

role_test_df["p_mcnemar_holm"] = holm_adjusted
role_test_df["significant_holm_05"] = role_test_df["p_mcnemar_holm"] < 0.05

# --- 3. Game-clustered bootstrap on the change in each role's share ----------
rng = np.random.default_rng(ROLE_SEED)

game_ids = slot_df["game_id"].unique()
game_index = {game_id: i for i, game_id in enumerate(game_ids)}
n_boot_games = len(game_ids)
role_index = {role: i for i, role in enumerate(roles)}

start_counts = np.zeros((n_boot_games, len(roles)))
end_counts = np.zeros((n_boot_games, len(roles)))
slot_counts = np.zeros(n_boot_games)

for game_id, start_role, end_role in zip(
    slot_df["game_id"], slot_df["start_role"], slot_df["end_role"]
):
    g = game_index[game_id]
    start_counts[g, role_index[start_role]] += 1
    end_counts[g, role_index[end_role]] += 1
    slot_counts[g] += 1

boot_idx = rng.integers(0, n_boot_games, size=(ROLE_N_BOOT, n_boot_games))
boot_start = start_counts[boot_idx].sum(axis=1)
boot_end = end_counts[boot_idx].sum(axis=1)
boot_slots = slot_counts[boot_idx].sum(axis=1)[:, None]

boot_delta = 100 * (boot_end / boot_slots - boot_start / boot_slots)
ci_low, ci_high = np.percentile(boot_delta, ROLE_CI, axis=0)

bootstrap_df = pd.DataFrame({
    "role": roles,
    "delta_percentage_points": [
        100 * ((slot_df["end_role"] == r).mean() - (slot_df["start_role"] == r).mean())
        for r in roles
    ],
    "ci_low": ci_low,
    "ci_high": ci_high,
})
bootstrap_df["ci_excludes_zero"] = (bootstrap_df["ci_low"] > 0) | (bootstrap_df["ci_high"] < 0)

role_test_df = role_test_df.merge(bootstrap_df, on="role")
role_test_df.to_csv(STATS_DIR / "role_distribution_significance.csv", index=False)

print("\n\n2-3. PER-ROLE TESTS  (McNemar + Holm, and game-clustered bootstrap CI)")
print("-" * 78)
display(role_test_df)

significant_roles = role_test_df.loc[role_test_df["significant_holm_05"], "role"].tolist()
print(
    "\nRoles whose share changes significantly (Holm-corrected McNemar, alpha = 0.05): "
    f"{significant_roles or 'none'}"
)

disagreements = role_test_df.loc[
    role_test_df["ci_excludes_zero"] & ~role_test_df["significant_holm_05"], "role"
].tolist()
print(
    "Roles where the bootstrap CI excludes zero but Holm-corrected McNemar does not: "
    f"{disagreements or 'none'}"
)
print("  (these are suggestive at best, and must not be reported as significant)")

print("\nSaved:", STATS_DIR / "role_distribution_significance.csv")

1. STUART-MAXWELL TEST OF MARGINAL HOMOGENEITY
------------------------------------------------------------------------------
categories tested: 14 (dropped, no discordant pairs: ['Moderator'])
chi2 = 20.45, df = 13, p = 0.0845
-> the start and end role distributions are NOT distinguishable at alpha = 0.05


2-3. PER-ROLE TESTS  (McNemar + Holm, and game-clustered bootstrap CI)
------------------------------------------------------------------------------


,role,n_start,n_end,n_lost,n_gained,p_mcnemar_raw,p_mcnemar_holm,significant_holm_05,delta_percentage_points,ci_low,ci_high,ci_excludes_zero
0,Drunk,45,13,45,13,0.000030,0.000452,True,-3.703704,-4.931297,-2.603320,True
1,Center Card,0,6,0,6,0.031250,0.437500,False,0.694444,0.116686,1.408492,True
2,Troublemaker,115,120,18,23,0.532709,1.000000,False,0.578704,0.000000,1.179245,False
3,Werewolf,221,229,65,73,0.551413,1.000000,False,0.925926,0.232829,1.635514,True
4,Seer,95,99,42,46,0.749329,1.000000,False,0.462963,-0.117374,1.142890,False
5,Villager,70,73,24,27,0.779768,1.000000,False,0.347222,0.000000,0.809249,False
6,Minion,27,29,9,11,0.823803,1.000000,False,0.231481,-0.222297,0.697674,False
7,Hunter,20,22,10,12,0.831812,1.000000,False,0.231481,0.000000,0.582072,False
8,Robber,118,120,104,106,0.945006,1.000000,False,0.231481,-0.574053,1.035703,False
9,Mason,12,11,6,5,1.000000,1.000000,False,-0.115741,-0.350467,0.000000,False



Roles whose share changes significantly (Holm-corrected McNemar, alpha = 0.05): ['Drunk']
Roles where the bootstrap CI excludes zero but Holm-corrected McNemar does not: ['Center Card', 'Werewolf']
  (these are suggestive at best, and must not be reported as significant)

Saved: C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\role_distribution_significance.csv


### How many Werewolves are in play

The base ONUW deck holds two Werewolf cards, and any number of them can sit in the centre,
so a game has 0, 1 or 2 Werewolves among the players. This matters for the outcome
analysis: with two Werewolves the wolves know each other and coordinate, with one the wolf
is alone, and with none there is no correct wolf-vote available at all.

In [23]:
werewolf_rows = []

for row in cleaned_index:
    werewolf_rows.append({
        "game_id": f"{row['source']} / {row['session_name']} / {row['game_key']}",
        "source": row["source"],
        "n_players": len(row["player_names"]),
        "n_werewolf_start": row["start_roles"].count("Werewolf"),
        "n_werewolf_end": row["end_roles"].count("Werewolf"),
        "n_minion_end": row["end_roles"].count("Minion"),
    })

werewolf_df = pd.DataFrame(werewolf_rows)
werewolf_df["werewolf_count_changed"] = (
    werewolf_df["n_werewolf_start"] != werewolf_df["n_werewolf_end"]
)
werewolf_df.to_csv(STATS_DIR / "werewolf_counts_per_game.csv", index=False)

n_games_total = len(werewolf_df)
n_slots = int(werewolf_df["n_players"].sum())

count_summary_df = pd.DataFrame({
    "n_games_start": werewolf_df["n_werewolf_start"].value_counts().sort_index(),
    "n_games_end": werewolf_df["n_werewolf_end"].value_counts().sort_index(),
}).fillna(0).astype(int)
count_summary_df["percent_start"] = (100 * count_summary_df["n_games_start"] / n_games_total).round(1)
count_summary_df["percent_end"] = (100 * count_summary_df["n_games_end"] / n_games_total).round(1)
count_summary_df.index.name = "n_werewolves"
count_summary_df.to_csv(STATS_DIR / "werewolf_count_distribution.csv")

print("Werewolves per game:")
display(count_summary_df)

print(f"Mean per game: start {werewolf_df['n_werewolf_start'].mean():.2f}, "
      f"end {werewolf_df['n_werewolf_end'].mean():.2f}")
print(f"Werewolf slots: start {int(werewolf_df['n_werewolf_start'].sum())} "
      f"({100 * werewolf_df['n_werewolf_start'].sum() / n_slots:.1f}% of {n_slots} player slots), "
      f"end {int(werewolf_df['n_werewolf_end'].sum())} "
      f"({100 * werewolf_df['n_werewolf_end'].sum() / n_slots:.1f}%)")

print()
print("Werewolves at voting time, by table size:")
display(pd.crosstab(werewolf_df["n_players"], werewolf_df["n_werewolf_end"], margins=True))

print("Werewolves at voting time, by source:")
display(pd.crosstab(werewolf_df["source"], werewolf_df["n_werewolf_end"], margins=True))

print("Start count -> end count:")
display(pd.crosstab(werewolf_df["n_werewolf_start"], werewolf_df["n_werewolf_end"]))

n_changed = int(werewolf_df["werewolf_count_changed"].sum())
n_minion_no_wolf = int(((werewolf_df["n_minion_end"] > 0) & (werewolf_df["n_werewolf_end"] == 0)).sum())
print(f"Games where the Werewolf count changed during the night: {n_changed}")
print(f"Games with a Minion but no Werewolf at voting time: {n_minion_no_wolf}")

print()
print("Saved:")
for filename in ["werewolf_counts_per_game.csv", "werewolf_count_distribution.csv"]:
    print("-", STATS_DIR / filename)

Werewolves per game:


,n_games_start,n_games_end,percent_start,percent_end
n_werewolves,,,,
0,28,26,14.7,13.6
1,105,101,55.0,52.9
2,58,64,30.4,33.5


Mean per game: start 1.16, end 1.20
Werewolf slots: start 221 (25.6% of 864 player slots), end 229 (26.5%)

Werewolves at voting time, by table size:


n_werewolf_end,0,1,2,All
n_players,,,,
3,2,3,3,8
4,17,53,30,100
5,5,32,21,58
6,2,13,10,25
All,26,101,64,191


Werewolves at voting time, by source:


n_werewolf_end,0,1,2,All
source,,,,
Ego4D,8,22,10,40
Youtube,18,79,54,151
All,26,101,64,191


Start count -> end count:


n_werewolf_end,0,1,2
n_werewolf_start,,,
0,25,3,0
1,1,98,6
2,0,0,58


Games where the Werewolf count changed during the night: 10
Games with a Minion but no Werewolf at voting time: 9

Saved:
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\werewolf_counts_per_game.csv
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\werewolf_count_distribution.csv


### Which roles are in play, and how often

The role counts above are over player slots. This cell is over **games**: in how many
games does each role appear at all, and with how many copies. ONUW is played with a deck
chosen per game rather than a fixed one, so the role frequencies are a property of what
these groups chose to play, not of a controlled design. The number of distinct deck
compositions is reported for that reason.

In [24]:
role_presence_rows = []

for role in roles:
    start_counts = [row["start_roles"].count(role) for row in cleaned_index]
    end_counts = [row["end_roles"].count(role) for row in cleaned_index]

    role_presence_rows.append({
        "role": role,
        "n_games_start": sum(1 for c in start_counts if c > 0),
        "percent_games_start": 100 * sum(1 for c in start_counts if c > 0) / n_games_total,
        "n_games_end": sum(1 for c in end_counts if c > 0),
        "percent_games_end": 100 * sum(1 for c in end_counts if c > 0) / n_games_total,
        "max_copies_in_a_game_start": max(start_counts),
        "max_copies_in_a_game_end": max(end_counts),
        "n_games_with_two_copies_start": sum(1 for c in start_counts if c == 2),
    })

role_presence_df = (
    pd.DataFrame(role_presence_rows)
    .sort_values("n_games_start", ascending=False)
    .reset_index(drop=True)
)
role_presence_df.to_csv(STATS_DIR / "role_presence_per_game.csv", index=False)

print(f"Role presence across the {n_games_total} games:")
display(role_presence_df)

# Deck composition: the multiset of starting roles actually dealt.
compositions = Counter(tuple(sorted(row["start_roles"])) for row in cleaned_index)

composition_df = (
    pd.DataFrame(
        [{"n_players": len(comp), "composition": ", ".join(comp), "n_games": n}
         for comp, n in compositions.items()]
    )
    .sort_values(["n_games", "n_players"], ascending=[False, True])
    .reset_index(drop=True)
)
composition_df.to_csv(STATS_DIR / "deck_compositions.csv", index=False)

print(f"Distinct starting-role compositions: {len(compositions)} across {n_games_total} games")
print()
print("Most common compositions:")
display(composition_df.head(8))

print("Distinct compositions by table size:")
display(
    composition_df.groupby("n_players")
    .agg(n_distinct_compositions=("composition", "nunique"), n_games=("n_games", "sum"))
)

print("Saved:")
for filename in ["role_presence_per_game.csv", "deck_compositions.csv"]:
    print("-", STATS_DIR / filename)

Role presence across the 191 games:


,role,n_games_start,percent_games_start,n_games_end,percent_games_end,max_copies_in_a_game_start,max_copies_in_a_game_end,n_games_with_two_copies_start
0,Werewolf,163,85.340314,165,86.387435,2,2,58
1,Robber,118,61.780105,120,62.827225,1,1,0
2,Troublemaker,115,60.209424,120,62.827225,1,1,0
3,Seer,95,49.738220,99,51.832461,1,1,0
4,Tanner,84,43.979058,84,43.979058,1,1,0
5,Villager,48,25.130890,51,26.701571,2,2,22
6,Insomniac,48,25.130890,49,25.654450,1,1,0
7,Drunk,45,23.560209,13,6.806283,1,1,0
8,Minion,27,14.136126,29,15.183246,1,1,0
9,Hunter,20,10.471204,22,11.518325,1,1,0


Distinct starting-role compositions: 127 across 191 games

Most common compositions:


,n_players,composition,n_games
0,4,"Robber, Troublemaker, Werewolf, Werewolf",8
1,4,"Robber, Tanner, Troublemaker, Werewolf",8
2,4,"Robber, Seer, Troublemaker, Werewolf",7
3,4,"Robber, Villager, Villager, Werewolf",4
4,4,"Seer, Tanner, Troublemaker, Werewolf",4
5,4,"Drunk, Robber, Troublemaker, Werewolf",4
6,4,"Robber, Seer, Tanner, Werewolf",3
7,4,"Robber, Seer, Villager, Werewolf",3


Distinct compositions by table size:


,n_distinct_compositions,n_games
n_players,,
3,7,8
4,50,100
5,47,58
6,23,25


Saved:
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\role_presence_per_game.csv
- C:\Users\annab\Documents\GitHub\masters_thesis_sdg\analysis\lai2023_corpus\tables\deck_compositions.csv
